src/util/tools.cc

```diff
@@ -230,23 +217,23 @@ bool isRegularFile(const fs::path& path, std::error_code& ec) noexcept
     ec.clear();
     return S_ISREG(statbuf.st_mode);
 #else
-    return fs::is_regular_file(path, ec);
+    return dirEnt.is_regular_file(ec);
 #endif
 }
 
-off_t getFileSize(const fs::path& path)
+off_t getFileSize(const fs::directory_entry& dirEnt)
 {
-    // unfortunately fs::file_size(path) is broken with old libstdc++ on 32bit systems (see #737)
+    // unfortunately fs::file_size() is broken with old libstdc++ on 32bit systems (see #737)
 #if defined(__GLIBCXX__) && (__GLIBCXX__ <= 20190406)
     struct stat statbuf;
-    int ret = stat(path.c_str(), &statbuf);
+    int ret = stat(dirEnt.path().c_str(), &statbuf);
     if (ret != 0) {
         throw_std_runtime_error("{}: {}", std::strerror(errno), path.c_str());
     }
 
     return statbuf.st_size;
 #else
-    return fs::file_size(path);
+    return dirEnt.file_size();
```

Comment:
What's the difference here?

Reply:
The scanning loop already has `directory_entry` and the constructor does the `stat`s. By using the method we avoid doing it all over again. The static method only takes `path`. That's the core of the cleanup: use `directory_entry` if already created.